### ISSUE TITLE
There are some missing items related to OAuth

### ISSUE THREAD

**Issue by ImanSharaf** [OWASP/ASVS#1431]
OAuth 2 related vulnerabilities are not uncommon because the OAuth specification is relatively vague and flexible by design. Also, OAuth bounties are usually highly paid (like [this one](https://ysamm.com/?p=763) with a $44000 bounty).
I believe we are missing a few checks for OAuth in the ASVS. I start with "Ensure that the application doesn't use any deprecated OAuth flows". We have two deprecated flows: 
1- [Implicit Flow](https://oauth.net/2/grant-types/implicit/)
2- [Resource Owner Password Credentials](https://oauth.net/2/grant-types/password/)

Furthermore, for SPAs (Single Page Applications) and mobile apps, the application should use PKCE flow. Also, I have seen customers that don't properly understand refresh token, access token and ID token and don't use them correctly. We should also mention that the refresh token should be stored in the cookie. I have some more items in my mind that I will add after I have heard your comments.

**Comment by tkisason** [comment]
Good idea, a common issue is the the need to add redirect_uri whitelist, since if an redirect_uri whitelist doesn't exist, it's possible to hijack accounts.

**Comment by elarlang** [comment]
You can propose more precise requirements with answering (first for yourself), what kind of problem or risk it addresses. We can help with potential duplicates (if the point is already covered in ASVS somehow).

> Good idea, a common issue is the the need to add redirect_uri whitelist, since if an redirect_uri whitelist doesn't exist, it's possible to hijack accounts.

There requirement for that:
_V5.1.5 Verify that URL redirects and forwards only allow destinations which appear on an allow list, or show a warning when redirecting to potentially untrusted content._

**Comment by ImanSharaf** [comment]
@elarlang what do you think about this one? 
"Ensure that the application doesn't use any deprecated OAuth flows". We have two deprecated flows:
1- [Implicit Flow](https://oauth.net/2/grant-types/implicit/)
2- [Resource Owner Password Credentials](https://oauth.net/2/grant-types/password/)

**Comment by elarlang** [comment]
With some exceptions, the style for requirement should be positive - what to achieve instead of what to not achieve.

What is the challenge here - if to fix the list to the standard, be future proof and past proof.
* fixed list - if we don't fix the list then it's "flowing requirement"
* future proof - what to do, if some flow allowed in standard is declared as deprecated
* past proof - we don't release ASVS too often, if some flow will be deprecated, then to cover it with up-to-date information, it requires some 5.0.* > 5.*.0 release

**Comment by ImanSharaf** [comment]
@tghosth @elarlang @jmanico What do you think about adding these new checks?
- Verify that the application implements only secure and recommended OAuth 2.0 flows, such as the Authorization Code Flow and Authorization Code Flow with PKCE. Ensure that deprecated flows, such as the Implicit Flow and Resource Owner Password Credentials Flow, are not used.
- Verify that the application securely validates redirect URIs during the OAuth 2.0 authorization process. Ensure that only registered and trusted redirect URIs are allowed, preventing potential open redirect vulnerabilities.
- Verify that the application generates, validates, and securely manages the OAuth 2.0 state parameter during the authorization process. Ensure that the state parameter is used to mitigate CSRF attacks and is generated securely, validated upon receiving the authorization code, and not reused.
- Verify that the application implements OAuth 2.0 Proof Key for Code Exchange (PKCE) for public clients, such as mobile or single-page applications. Ensure that PKCE enhances the security of the authorization code flow by mitigating the risk of authorization code interception.

**Comment by jmanico** [comment]
These look solid

1) you need the state parameter to address CSRF OR PKCE but not both.

2) resource owner client credentials is ok when the client and auth server are the same vendor; but your point is taken well, I get it

3) PKCE is important even for confidential clients!

--
Jim Manico
@Manicode

**Comment by elarlang** [comment]
Please take a look, maybe this topics are covered in https://github.com/OWASP/ASVS/issues/996

**Comment by tghosth** [comment]
Agree with @elarlang. We even have a PR with some proposed changes in #1494. Please check if anything you have mentioned is not in that issue or PR and comment there. I am going to close this as a duplicate for now. I would particiularly be interested in input on how we can focus down the fewest important requirements as I believe the proposed section may currently be too large and complicated.
